Refactor of WebView URL Loading (#3061)
OWASP/owasp-mastg: tests/android/MASVS-CODE/MASTG-TEST-0x27-2.md

---
Title: Testing for URL Loading in WebViews
ID: MASTG-TEST-0x27-2
Link: https://mas.owasp.org/MASTG/tests/android/MASVS-CODE/MASTG-TEST-0027/
Platform: android
type: [dynamic]
MASVS v1: ['MSTG-PLATFORM-2']
MASVS v2: ['MASVS-CODE-4']
---

## Overview

> By default, navigation events inside of a WebView will redirect to the default browser application. However, it is possible to stay within the WebView and handle all new page loads. This can be dangerous, as the new page may be malicious and interact with either the JavaScript bridge, or phish the user. The application should monitor navigation events inside the WebView to make sure that only legitimate pages are loaded, while others are redirected to the browser application.

Reviewer: ```suggestion
By default, navigation events inside of a WebView will redirect to the default browser application. However, it is possible to stay within the WebView and handle all new page loads. This can be dangerous, as the new page may be malicious and interact with either the JavaScript bridge (see @MASWE-0068), or phish the user. The application should monitor navigation events inside the WebView to make sure that only legitimate pages are loaded, while others are redirected to the browser application.
```

## Steps

1. Launch the application and make sure you can hook functions (see @MASTG-TECH-0043).
2. Hook the following functions to see if they are executed:

>    1. WebViewClient.shouldOverrideUrlLoading
>    2. WebViewClient.shouldInterceptRequest
>    3. WebSettings.setSafeBrowsingEnabled

Reviewer: ```suggestion
   1. `WebViewClient.shouldOverrideUrlLoading`
   2. `WebViewClient.shouldInterceptRequest`
   3. `WebSettings.setSafeBrowsingEnabled`
```

> 3. Use any WebView inside the app and trigger navigation events

Reviewer: ```suggestion
3. Use every WebView inside the app and trigger navigation events, using a variety of trusted and non-trusted URLs
```

Reviewer: Not sure if it is obvious what a non-trusted URL is in this context? But I think a tester needs to try various URLs, otherwise, potentially only the happy path for the Custom Webview will be triggered.

Reviewer 2: If our goal is "simply" to know if the app uses these functions, wouldn't it be sufficient to ensure we trigger every WebView?

We could force them to load untrusted URLs, maybe using Frida and overriding the URL to be loaded, but isn't that maybe too much for the scope of this test?
